### PR TITLE
Fix text being bumped out of codeblock.

### PR DESCRIPTION
**Pull request by rjb25**
Related to this [issue](https://github.com/yogthos/markdown-clj/issues/120)
I know you said that you wanted to look at it as it could have adverse affects, but as far as I can tell this is simply fixing a typo. Content that was before \`\`\` should not suddenly be after the \`\`\`.

### CODE REVIEW

**Comment by rjb25** [comment]
That test failed hard... How with such a small change?

**Comment by rjb25** [comment]
This second pull request is a working fix for the code block issues. You may be able to find a more elegant solution. The first pull requests content was removed from this pull request. If you decide to use the content from the first pull request you would have to update your tests and then copy that one line change in.

File: src/cljc/markdown/transformers.cljc [hunk 1]
```diff
@@ -125,9 +125,9 @@
     (str " " text) text))
 
 (defn paragraph
-  [text {:keys [eof heading inline-heading temp hr code lists blockquote paragraph last-line-empty?] :as state}]
+  [text {:keys [eof heading inline-heading temp hr code codeblock lists blockquote paragraph last-line-empty? end-codeblock?] :as state}]
   (cond
-    (and paragraph lists)
+    (or (and paragraph lists) (and paragraph codeblock))
```
Comment (yogthos) [inline src/cljc/markdown/transformers.cljc]: this shouldn't be a valid case, the paragraph and codeblock are mutually exclusive states

File: src/cljc/markdown/transformers.cljc [hunk 2]
```diff
@@ -138,11 +138,11 @@
       [(str (paragraph-text last-line-empty? text) "</p>") (assoc state :paragraph false)]
       [(paragraph-text last-line-empty? text) state])
 
-    (and (not eof) (not (string/blank? text)) (or (:inline-heading temp) last-line-empty?))
-    [(str "<p>" text) (assoc state :paragraph true :last-line-empty? false)]
+    (and (not eof) (not (string/blank? text)) (or (:inline-heading temp) last-line-empty? (= end-codeblock? "now")))
```
Comment (yogthos) [inline src/cljc/markdown/transformers.cljc]: using a keyword would be more idiomatic than a string here

**Comment by yogthos** [comment]
unfortunately the pr doesn't actually address the problem, for example the following

    "```\nfoo```\nbar"

is rendered as

    <pre><code></code></pre>foo<p>bar</p>

**Comment by yogthos** [comment]
As I explained in the issue, the fix is a bit trickier. The trim behavior is not a typo, but is there to handle the case where next line content has already been prepended to the  \`\`\` string generated at the end of the code block.

File: src/cljc/markdown/transformers.cljc (hunk 1 above)
Comment (rjb25) [inline src/cljc/markdown/transformers.cljc]: However codeblocks are not actually mutually exclusive from paragraphs. as seen here:
`user=> (md-to-html-string "hey\n```\nhey\n```\nhey")
"<p>hey<pre><code>hey\n </code></pre> hey</p>"`
Note the paragraph is ending after the codeblock. This is what is causing the codeblock to take up an extra line. The above code says *if you are in a paragraph while in a codeblock, end the paragraph.* The rest of the code solves the issue of making sure the paragraph starts back up again after the codeblock.

Comment (yogthos) [inline src/cljc/markdown/transformers.cljc]: That should actually get split up, e.g:

    (md-to-html-string "hey\n\n    hey\n\nhey")

will currently produce:

    <p>hey</p><pre><code>hey</code></pre><p>hey</p>

Comment (rjb25) [inline src/cljc/markdown/transformers.cljc]: Where are the backticks?

Comment (yogthos) [inline src/cljc/markdown/transformers.cljc]: The alternative way to specify a code block is to use 4 spaces

Comment (yogthos) [inline src/cljc/markdown/transformers.cljc]: Using the backticks you'll still get content that's broken up:

    "hey\n\n```\nhey\n```\n\nhey"

    <p>hey</p><pre><code>hey
    </code></pre><p>hey</p>

Comment (rjb25) [inline src/cljc/markdown/transformers.cljc]: Yes I agree it should get split up. My pull request makes the previous example I gave give this:
```
user=> (md-to-html-string "hey\n```\nhey\n```\nhey\nhey")
"<p>hey</p><pre><code>hey\n</code></pre><p>hey hey</p>"
```

Comment (yogthos) [inline src/cljc/markdown/transformers.cljc]: Right, but I don't think it solves the underlying problem. I'll have to think about this a bit. Thanks for taking a look.

**Comment by rjb25** [comment]
In response to "unfortunately the pr doesn't actually address the problem" It does. However I removed the part that fixed that particular issue from the latest pull request to avoid failing one of the tests and to keep the two fixes separate. When including the code from the first request with this request you example:
```
"```\nfoo```\nbar"
```
Is redered as
```
"<pre><code>foo</code></pre><p>bar</p>"
```

File: src/cljc/markdown/transformers.cljc (hunk 1 above)
Comment (rjb25) [inline src/cljc/markdown/transformers.cljc]: Ok, hopefully this has helped!

**Comment by yogthos** [comment]
Right, but the code would have to be fixed first to avoid failing the test.

**Comment by rjb25** [comment]
One thing you might want to look at. In your "Using the backticks you'll still get content that's broken up:" comment the content is not actually broken up. There is one hey before the code block, one hey inside and one after as desired. If you are refering to the newline, that is ok. If there is a newline with a space after it though, then the additional space is rendered and there is a problem.

**Comment by yogthos** [comment]
so [here's](https://github.com/yogthos/markdown-clj/commit/435fc0a08c36bfacffa4ff62b4fe6ee86d876feb) the fix for handling new line after the code block, and text on the same line as \`\`\`.

**Pull request closed without merging**